tests.py

```diff
@@ -0,0 +1,303 @@
+    client.rm('/', recursive=True)
+
+
+@mkdirs_recursively(
+    tree={
+        'foo': {
+            'bar': {
+                'baz': None
+            }
+        }
+    }
+)
+def test_exists(client):
+    assert client.exists('/foo/')
+    assert not client.exists('/foo')
+    assert client.exists('/foo//')
+    assert client.exists('/foo/bar/')
+    assert client.exists('/foo/bar/baz')
+    assert not client.exists('/foo/bar/baz/')
+
+
+@pytest.mark.skip()
+@mkdirs_recursively(
+    tree={
+        'foo': {
+            'bar': None
+        }
+    }
+)
+def test_put_data(client):
+    assert client.exists('/foo/bar')
+
+
+@pytest.mark.skip()
+@mkdirs_recursively(
+    tree={
+        'foo': {
+            'bar': None
+        }
+    }
+)
+def test_put_file(client):
+    assert client.exists('/foo/bar')
+
+
+@mkdirs_recursively(
+    tree={
+        'foo1': {
+            'bar': [],
+        },
+        'foo2': {
+            'bar': {
+                'baz': None,
+            }
+        },
+        'foo3': {
+            'bar': {
+                'baz1': [],
+                'baz2': [],
+            }
+        },
+    }
+)
+def test_listdir(client):
+    assert client.listdir('/').sort() == ['foo1/', 'foo2/', 'foo3/'].sort()
+    assert client.listdir('/foo1/') == ['bar/']
+    assert client.listdir('/foo2/') == ['bar/']
+    assert client.listdir('/foo2/bar/') == ['baz']
+    assert client.listdir('/foo3/bar/').sort() == ['baz1/', 'baz2/'].sort()
+
+
+@mkdirs_recursively(
+    tree={
+        'foo1': {
+            'bar': [],
+        },
+        'foo2': {
+            'bar': {
+                'baz1': [],
+                'baz2': None,
+            }
+        },
+        'foo3': {
+            'bar': {
+                'baz1': [],
+                'baz2': [],
+            }
+        }
+    }
+)
+def test_listdir_only_files(client):
+    assert client.listdir('/', only_files=True) == []
+    assert client.listdir('/foo1/', only_files=True) == []
+    assert client.listdir('/foo2/', only_files=True) == []
+    assert client.listdir('/foo2/bar/', only_files=True) == ['baz2']
+    assert client.listdir('/foo3/bar/', only_files=True) == []
+
+
+@mkdirs_recursively(
+    tree={
+        'foo': {
+            'bar': {
+                'baz': None
+            }
+        }
+    }
+)
+def test_isdir(client):
+    assert client.isdir('/foo/')
+    assert client.isdir('/foo/bar/')
+    assert not client.isdir('/foo/bar/baz')
+
+
+@mkdirs_recursively(
+    tree={
+        'foo1': {
+            'bar': [],
+        },
+        'foo2': {
+            'bar': {
+                'baz1': [],
+                'baz2': None,
+            }
+        },
+        'foo3': {
+            'bar': {
+                'baz1': [],
+                'baz2': [],
+            }
+        }
+    }
+)
+def test_rmdir(client):
+    client.rmdir('/foo1/bar/')
+    assert not client.exists('/foo1/bar/')
+
+    pytest.raises(ValueError, client.rmdir, '/foo2/bar/baz2')
```

Comment:
```suggestion
    with pytest.raises(ValueError):
        client.rmdir('/foo2/bar/baz2')
```
You can also validate the error message if you want to